mailbox/plugin/deleted-messages-vault-blobstore-memory/src/main/java/org/apache/james/vault/MemoryDeletedMessageMetadataVault.java

```diff
@@ -19,40 +19,78 @@
+    private final Table<BucketName, User, Map<MessageId, DeletedMessageWithStorageInformation>> table;
+
+    public MemoryDeletedMessageMetadataVault() {
+        table = HashBasedTable.create();
+    }
+
     @Override
-    public Publisher<Void> store(DeletedMessageWithStorageInformation deletedMessage) {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<Void> store(DeletedMessageWithStorageInformation deletedMessage) {
+        BucketName bucketName = deletedMessage.getStorageInformation().getBucketName();
+        User owner = deletedMessage.getDeletedmessage().getOwner();
+        MessageId messageId = deletedMessage.getDeletedmessage().getMessageId();
+
+        Map<MessageId, DeletedMessageWithStorageInformation> userVault = userVault(bucketName, owner);
+        userVault.put(messageId, deletedMessage);
+        table.put(bucketName, owner, userVault);
+
+        return Mono.empty();
     }
 
     @Override
-    public Publisher<Void> removeBucket(BucketName bucketName) {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<Void> removeBucket(BucketName bucketName) {
+        table.row(bucketName).clear();
+
+        return Mono.empty();
     }
 
     @Override
-    public Publisher<Void> remove(BucketName bucketName, User user, MessageId messageId) {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<Void> remove(BucketName bucketName, User user, MessageId messageId) {
+        userVault(bucketName, user).remove(messageId);
+
+        return Mono.empty();
     }
 
     @Override
-    public Publisher<StorageInformation> retrieveStorageInformation(User user, MessageId messageId) {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<StorageInformation> retrieveStorageInformation(User user, MessageId messageId) {
+        return Flux.from(listBuckets())
+            .flatMap(bucket -> Mono.justOrEmpty(userVault(bucket, user).get(messageId)))
+            .map(DeletedMessageWithStorageInformation::getStorageInformation)
+            .next();
     }
 
     @Override
-    public Publisher<DeletedMessageWithStorageInformation> listMessages(BucketName bucketName, User user) {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<DeletedMessageWithStorageInformation> listMessages(BucketName bucketName, User user) {
+        return Flux.fromIterable(Optional.ofNullable(table.get(bucketName, user))
+            .map(Map::values)
+            .orElse(ImmutableList.of()));
     }
 
     @Override
-    public Publisher<BucketName> listBuckets() {
-        throw new NotImplementedException("Not yet implemented");
+    public synchronized Publisher<BucketName> listBuckets() {
+        return Flux.fromIterable(table.rowKeySet());
```

Comment:
all the `synchronized` are wrong: you don't protect the collection during its usage. BTW, that would be a problem to lock the collection in a Publisher because you can never be sure it will be consumed.

Reply:
> BTW, that would be a problem to lock the collection in a Publisher because you can never be sure it will be consumed.

What do you mean? No `.block()` -> no code execution -> no need of locks...